tests/run-make/pgo-branch-weights/rmake.rs

```diff
@@ -0,0 +1,67 @@
+// This test generates an instrumented binary - a program which
+// will keep track of how many times it calls each function, a useful
+// feature for optimization. Then, an argument (aaaaaaaaaaaa2bbbbbbbbbbbb2bbbbbbbbbbbbbbbbcc)
+// is passed into the instrumented binary, which should react with a number of function calls
+// fully known in advance. (For example, the letter 'a' results in calling f1())
+
+// If the test passes, the expected function call count was added to the use-phase LLVM-IR.
+// See https://github.com/rust-lang/rust/pull/66631
+
+// FIXME(mati865): MinGW GCC miscompiles compiler-rt profiling library but with Clang it works
+// properly. Since we only have GCC on the CI ignore the test for now.
+
+//@ needs-profiler-support
+//@ ignore-windows-gnu
+//@ ignore-cross-compile
+
+use run_make_support::{llvm_profdata, rustc, rustdoc, stdin_command, target, tmp_dir};
+
+fn main() {
+    // For some very small programs GNU ld seems to not properly handle
+    // instrumentation sections correctly. Neither Gold nor LLD have that
+    // problem.
+    let link_args = (cfg!(target_os = "linux") && target().contains("x86"))
+        .then_some("-Clink-args=-fuse-ld=gold");
+
+    let path_prof_data_dir = tmp_dir().join("prof_data_dir");
+    let path_merged_profdata = path_prof_data_dir.join("merged.profdata");
+    let invoc = rustc().input("opaque.rs");
+    if let Some(link_args) = link_args {
+        invoc.arg(link_args);
+    }
+    invoc.run();
+    fs::create_dir_all(&path_prof_data_dir);
+    let invoc = rustc()
+        .input("interesting.rs")
+        .profile_generate(&path_prof_data_dir)
+        .opt()
+        .codegen_units(1);
+    if let Some(link_args) = link_args {
+        invoc.arg(link_args);
+    }
+    invoc.run();
+    let invoc = rustc().input("main.rs").profile_generate(&path_prof_data_dir).opt();
+    if let Some(link_args) = link_args {
+        invoc.arg(link_args);
+    }
+    invoc.run();
+    run("main aaaaaaaaaaaa2bbbbbbbbbbbb2bbbbbbbbbbbbbbbbcc");
+    llvm_profdata()
+        .merge()
+        .output(&path_merged_profdata)
+        .input(path_prof_data_dir)
+        .command_output();
+    let invoc = rustc()
+        .input("interesting.rs")
+        .profile_use(path_merged_profdata)
+        .opt()
+        .codegen_units(1)
+        .emit("llvm-ir");
+    if let Some(link_args) = link_args {
+        invoc.arg(link_args);
+    }
+    invoc.run();
+
+    let interesting_ll = tmp_dir().join("interesting.ll");
+    stdin_command(env::var("LLVM_FILECHECK"), interesting_ll, "filecheck-patterns.txt");
```

Comment:
This invocation is wonky. I wonder if we should just have a `llvm_filecheck` command wrapper that has an API like (names and exact design subject to improvements):

```rs
struct LlvmFilecheck { cmd: Command, ... }

impl LlvmFilecheck {
    fn new(&mut self, buf: &[u8]) -> Self {
        let path = env::var("LLVM_FILECHECK").expect("LLVM_FILECHECK env var not specified");
        let cmd = Command::new(path);
        Self { cmd, ... } 
    }
    fn stdin(&mut self, buf: &[u8]) -> &mut Self {}
    fn input<P: AsRef<Path>>(&mut self, path: P) -> &mut Self {}
    fn patterns<P: AsRef<Path>>(&mut self, path: P) -> &mut Self {}
}
```

Reply:
Implemented. I took inspiration from `rustc.rs` in `run-make-support` which has a `command_output` function. Juggling around with the types may cause a CI fail, I will see what happens.